tests/run/test_named_expressions.py

```diff
@@ -1,119 +1,169 @@
 # copied from cpython with minimal modifications (mainly exec->cython.inline, and a exception strings)
+# cython: language_level=3
 
 import os
 import unittest
 import cython
-from Cython.Build.Inline import cython_inline
 from Cython.Compiler.Main import CompileError
+from Cython.Build.Inline import cython_inline
+import sys
 
 if cython.compiled:
-    def exec_(code, locals_, globals_):
+    class StdErrHider:
+        def __enter__(self):
+            try:
+                from StringIO import StringIO
+            except ImportError:
+                from io import StringIO
+
+            self.old_stderr = sys.stderr
+            self.new_stderr = StringIO()
+            sys.stderr = self.new_stderr
+
+            return self
+
+        def __exit__(self, exc_type, exc_value, traceback):
+            sys.stderr = self.old_stderr
+
+        @property
+        def stderr_contents(self):
+            return self.new_stderr.getvalue()
+
+    def exec(code, globals_=None, locals_=None):
```

Comment:
Wasn't the `exec()` implementation in the other test modules (like `test_asyncgen.py` or `test_coroutines_pep492.py`) enough here? Maybe we could extract that into the test runner and inject it into the test module's namespace there (or the builtins) based on a test tag, so that the tests can just call it from there.

Reply:
I did start with copying exec from other test modules. (I think I used `test_grammar` as the basis). The issues were:

* The `cython_inline` runs in the a function scope (`__invoke`) which returns `locals()`, but some of the tests here actually explicitly use `global` so I need to get the real globals out of it too.
* A lot of the tests here are for compilation failures. `cython_inline` raises parser failures as  an exception with the correct error message, but any failure after the parser gets raised with the filename as the error message.

It really did need most of the modifications here 

I could see the value of a shared `cython_exec` though that behaves as close to the "real" `exec` as is possible (mainly for testing, but it could also be exposed)